### DISCUSSION TITLE
Preview draft content from CMSs in static sites (think: Next.js Preview Mode)

### DISCUSSION THREAD

**Discussion by angeloashmore** [withastro/roadmap#564] [Proposal]
# Summary

Astro is primarily a static site generator (SSG). SSG works great for production builds, but does not work well for previewing draft content before publishing. Astro could overcome this limitation by offering a way to dynamically switch between SSG and SSR modes _at runtime,_ which would retain great production performance and enable content writers to preview draft content.

# Background & Motivation

Previewing draft content from a content management system (CMS) is a core part of a content writer's workflow. Being able to see how content renders on a site _before_ publishing allows writers to verify their content looks correct in its final context and saves them from embarrassing publication mistakes.

Other frameworks like Next.js and Nuxt provide this functionality, usually specifically to cater to CMSs.

# Goals

- Support for a "preview" mode of your Astro build using server-rendering, regardless of the deployment adapter used for production (SSG or SSR)
- Ability to run concurrent previews scoped to the initiating previewer

# Example

- Next.js provides this capability via [Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode)
- Nuxt provides this capability via [Preview Mode](https://nuxtjs.org/docs/features/live-preview/)

**Comment by natemoo-re** [comment]
Yep, having an official preview mode would be incredibly useful!

For now, we've done a bit of work to make [Hybrid Rendering](https://docs.astro.build/en/guides/server-side-rendering/#hybrid-rendering) support more dynamic. Currently `export const prerender = import.meta.env.PRERENDER_ENABLED` should be supported for some use cases, but this is still much more manual than a proper preview mode would be.

Relevant PRs:
- https://github.com/withastro/astro/pull/6772
- https://github.com/withastro/astro/pull/6910

**Comment by FredKSchott** [comment]
> Support for a "preview" mode of your Astro build using server-rendering, regardless of the deployment adapter used for production (SSG or SSR)

Is this a reasonable goal? I can't imagine how live preview could work for any SSG site, and I've never heard of one support this since an SSG build is by definition a bunch of static already-built assets. If anyone has an example of this I'd love to see it!

If #539 would solve this, then I agree this is something already on the roadmap for this quarter that we shouldn't let drop!

**Reply by angeloashmore** [reply]
I suppose there's potential to run a component's server-side code on the client conditionally, but there are many security and technical concerns that would make it not feasible.

Should the goal be updated to reduce the scope to only SSR-capable adapters? We would like all pages to be rendered statically whenever possible (enabled by hybrid rendering), but with the capability of switching any page to SSR at runtime based on some conditions, like the presence of a cookie or URL param.

**Reply by jasikpark** [reply]
That sounds like ISR honestly "cache and prerender the page unless these cookies or url param change"

**Reply by bholmesdev** [reply]
I think this is a reasonable goal, though maybe we could clarify a bit? We want to find an easy way to generate an SSR build for previews, and an SSG build for production. This would require 2 separate Astro configs per-environment to apply the appropriate adapter, which we don't have a best practice for today.

**Reply by FredKSchott** [reply]
> We want to find an easy way to generate an SSR build for previews, and an SSG build for production.

Oh yea, this is something @tony-sull and I have actually discussed! It wasn't going to be a priority until next quarter, but if teams are asking for this it could be worth exploring sooner. I understand how this could solve for that goal as written, thanks for helping to clarify!

**Reply by Natetronn** [reply]
To add context and demonstrate a specific use case, Directus has an article on setting up Live Preview for both Next and Nuxt:

https://docs.directus.io/guides/headless-cms/live-preview/

Reading the Next article really drives home why their "Live Preview" is an important feature to have for content editors and demonstrates very well how such a feature (Preview Mode along with Live Preview) works in practice; if Preview Mode was implemented in Astro (in theory).

And of course, I'm curious if any progress been made on this front?

**Comment by Perni1984** [comment]
We're actively considering Astro for our next website projects and I wonder if this would be possible already today? E.g. setup an environment called preview.mydomain.com with an SSR astro config and an environment called www.mydomain.com with SSG build for production? Although no best practices exist (yet) I think it should be possible to set this up manually, right?

From what I get here https://docs.astro.build/en/guides/configuring-astro/#environment-variables - we could event control SSR vs. SSG via an environment variable.

**Reply by delucis** [reply]
100% — I believe this is what some people already do.

**Reply by andrejilderda** [reply]
Yes indeed, this is how we're handling previews also.

**Reply by Natetronn** [reply]
This is what I've been trying to get worked out as well. There is an article about it using Statamic as the preview CMS, but I've been trying to do it with Directus and fell short.

https://maciekpalmowski.dev/implementing-live-previews-in-headless-statamic-when-using-astro/

I'll give it another go soon, but in the mean time, would either of you be willing to elaborate on how you achieved this or perhaps be willing to share a quick gist or repo, by chance?

**Reply by andrejilderda** [reply]
Let me be clear, what we're talking about here is NOT a live preview as shown in the linked article. This means that you don't see the changes as you type, but you need to hit the Save-button before being able to see the changes (which works fine for us because our CMS distinguishes between saving and publishing). For this we have a preview environment setup where we use `output: 'server'`. For production we generate static pages (`output: 'static'`).

This discussion however is how Astro would be able to handle live previews. This is something Astro is not capable of yet.

**Reply by palmiak** [reply]
As an author of this article, I want to clarify that from the Statamic perspective, those are live previews. 

When you type something, it refreshes the iframe with the latest content from the API. But of course, it being "live" is an illusion created by the CMS itself.

**Reply by Natetronn** [reply]
@palmiak Thanks for sharing that article, btw! 

Hopefully this gets ironed out on the Astro side soon, to make it easier for all CMS and Frameworks to have Live Previews WITHOUT having to resort to creating a SSR clone of the original.

**Reply by andrejilderda** [reply]
@palmiak Ah, yes of course. My reference was Next.js which can update without doing full page reload.

**Comment by kb1995** [comment]
Any updates on this? I'd love to start using Astro for my client projects and I think this is the last blocker :)

**Comment by stefanprobst** [comment]
in next.js, "preview mode" has been superseded by ["draft mode"](https://nextjs.org/docs/app/building-your-application/configuring/draft-mode).

this works really nice for real-time previews like with [keystatic](https://keystatic.com/docs/recipes/real-time-previews). would be very cool if this could work with Astro as well!

**Comment by bholmesdev** [comment]
Well, I'm happy to share that **[ISR mode is now implemented](https://docs.astro.build/en/guides/integrations-guide/vercel/#isr) on the Vercel adapter!** We also [have a reference on bypassing the ISR cache](https://docs.astro.build/en/guides/integrations-guide/vercel/#excluding-paths-from-caching) for previewing CMS changes. This is what Vercel calls "draft mode." However, I'll admit the docs are early. If anyone wants to try and offer feedback, please do

**Reply by codejagaban** [reply]
I am trying this out today🚀

**Reply by lilnasy** [reply]
@codejagaban Vercel hasn't updated the [Astro doc](https://vercel.com/docs/frameworks/astro) yet, but the instructions on the [SvelteKit doc](https://vercel.com/docs/frameworks/sveltekit#draft-mode) would be pretty close!

**Reply by ducan-ne** [reply]
@bholmesdev is this support with other adapters like cloudflare

**Reply by giorgiofnool** [reply]
Is there a way to implement this in a self-hosted node server? Or a guide to follow? Thanks
